intbot/core/integrations/github.py

```diff
@@ -55,127 +77,175 @@ def as_discord_message(self):
         return f"[{self.title}]({self.url})"
 
 
-@dataclasses.dataclass
-class GithubDraftIssue:
+class GithubDraftIssue(BaseModel):
     id: str
     title: str
 
     def as_discord_message(self):
         return self.title
 
 
-def fetch_github_item_details(item_id):
-    headers = {
-        "Authorization": f"Bearer {settings.GITHUB_API_TOKEN}",
-        "Content-Type": "application/json",
-    }
-    payload = {"query": query, "variables": {"itemId": item_id}}
-    response = httpx.post(GITHUB_API_URL, json=payload, headers=headers)
-    if response.status_code == 200:
-        return response.json()["data"]["node"]["content"]
-    else:
-        raise Exception(f"GitHub API error: {response.status_code} - {response.text}")
+class GithubWebhook:
+    """
+    Base class for all the other specific types of webhooks.
+    """
 
+    def __init__(self, action: str, headers: dict, content: dict, extra: dict):
+        self.action = action
+        self.headers = headers
+        self.content = content
+        self.extra = extra
+        self._project = None
+        self._repository = None
+
+    @classmethod
+    def from_webhook(cls, wh: Webhook):
+        assert wh.extra, "Extra should be set already at this point"
+        return cls(
+            content=wh.content,
+            headers=wh.meta,
+            extra=wh.extra,
+            action=wh.event,
+        )
 
-CONTENT_TYPE_MAP = {
-    "Issue": GithubIssue,
-    "DraftIssue": GithubDraftIssue,
-}
+    def get_project(self):  # pragma: no cover
+        """Used in the discord channel router"""
+        raise NotImplementedError("Implement in child class")
+
+    def get_repository(self):  # pragma: no cover
+        """Used in the discord channel router"""
+        raise NotImplementedError("Implement in child class")
 
 
-class GithubProjectV2Item:
+class GithubProjectV2Item(GithubWebhook):
     # NOTE: This might be something for pydantic schemas in the future
 
-    def __init__(self, content: dict):
-        self.content = content
+    @property
+    def sender(self):
+        sender = self.get_sender()
 
-    def action(self):
-        if self.content["action"] == "edited":
-            action = "changed"
-        elif self.content["action"] == "created":
-            action = "created"
-        else:
-            raise ValueError(f"Action unsupported {self.content['action']}")
+        return f"[@{sender.login}]({sender.html_url})"
 
-        return action
+    def get_sender(self) -> GithubSender:
+        return GithubSender.model_validate(self.content["sender"])
 
-    def sender(self):
-        login = self.content["sender"]["login"]
-        url = self.content["sender"]["html_url"]
+    def github_object(self) -> GithubDraftIssue | GithubIssue:
+        content = self.extra["content"]
+        typename = content.pop("__typename")
```

Comment:
Why pop and not `content.get()`? With pop i assume it's not needed anymore, but I don't see that there is a serialization of the whole object (minus typename) afterwards or so. I didn't open it in IDE, so there might be - if so, then feel free to disregard this comment.

Reply:
yes, __typename is no longer needed - it's extra information to help us choose a schema.